src/test/java/picard/analysis/replicates/CollectIndependentReplicatesMetricTest.java

```diff
@@ -208,15 +209,19 @@ public void simpleTest(final String vcf, final String bam, final Map<String, Obj
      *
      */
     private static File convertSamToBam(final String sam) throws IOException {
-        final MergeSamFiles msf = new MergeSamFiles();
-        final File bam = new File(bamOutDir, sam.replaceAll("sam$", "bam"));
-        final int returnCode = msf.instanceMain(
-                new String[]{
-                        "INPUT=" + (new File(testdir, sam).getAbsolutePath()),
-                        "CREATE_INDEX=true",
-                        "OUTPUT=" + bam.getAbsolutePath()});
-        Assert.assertEquals(returnCode, 0);
-
-        return bam;
+        try {
+            final MergeSamFiles msf = new MergeSamFiles();
+            final File bam = new File(bamOutDir, sam.replaceAll("sam$", "bam"));
+            final int returnCode = msf.instanceMain(
+                    new String[]{
+                            "INPUT=" + (new File(testdir, sam).getAbsolutePath()),
+                            "CREATE_INDEX=true",
+                            "OUTPUT=" + bam.getAbsolutePath()});
+            Assert.assertEquals(returnCode, 0);
+            return bam;
+        } finally {
+            // reset the global defaults to eliminate side effects in downstream code
+            SAMFileWriterFactory.setDefaultCreateIndexWhileWriting(false);
```

Comment:
hmmm. this is quite insidious....how can we make sure that it doesn't happen again?

Reply:
Not sure, short of fixing https://github.com/broadinstitute/picard/issues/1342. One other option I considered was to change `CommandLineProgram` so that instead of [this](https://github.com/broadinstitute/picard/blob/master/src/main/java/picard/cmdline/CommandLineProgram.java#L243), it just does this:

`SAMFileWriterFactory.setDefaultCreateIndexWhileWriting(CREATE_INDEX);`
 
But that seems just as sketchy, and overwrites the value even in the case where it was set intentionally via a property or whatever. And there are several such options (IIRC VariantContextWriterBuilder has some of these too).

Reply:
There already exist other non-test code paths that can leave these side effects, such as just running SortVcfs or MergeVcfs with default args, since they default to CREATE_INDEX=true.